pubsub/iterator.go

```diff
@@ -427,127 +437,111 @@ func (it *messageIterator) handleKeepAlives() {
-	// Account for the Subscription and AckDeadlineSeconds fields.
-	overhead := calcFieldSizeString(it.subName) + calcFieldSizeInt(int(deadlineSec))
-	return it.sendAckIDRPC(m, maxPayload-overhead, func(ids []string) error {
+
+	ackIDs := make([]string, 0, len(m))
+	for k := range m {
+		ackIDs = append(ackIDs, k)
+	}
+	var toSend []string
+	for len(ackIDs) > 0 {
+		toSend, ackIDs = splitRequestIDs(ackIDs, ackIDBatchSize)
 		if deadline == 0 {
-			recordStat(it.ctx, NackCount, int64(len(ids)))
+			recordStat(it.ctx, NackCount, int64(len(toSend)))
 		} else {
-			recordStat(it.ctx, ModAckCount, int64(len(ids)))
-		}
-		addModAcks(ids, deadlineSec)
-		// Retry this RPC on Unavailable for a short amount of time, then give up
-		// without returning a fatal error. The utility of this RPC is by nature
-		// transient (since the deadline is relative to the current time) and it
-		// isn't crucial for correctness (since expired messages will just be
-		// resent).
-		cctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-		defer cancel()
-		bo := gax.Backoff{
-			Initial:    100 * time.Millisecond,
-			Max:        time.Second,
-			Multiplier: 2,
+			recordStat(it.ctx, ModAckCount, int64(len(toSend)))
 		}
-		for {
-			err := it.subc.ModifyAckDeadline(cctx, &pb.ModifyAckDeadlineRequest{
-				Subscription:       it.subName,
-				AckDeadlineSeconds: deadlineSec,
-				AckIds:             ids,
-			})
-			switch status.Code(err) {
-			case codes.Unavailable:
-				if err := gax.Sleep(cctx, bo.Pause()); err == nil {
-					continue
-				}
-				// Treat sleep timeout like RPC timeout.
-				fallthrough
-			case codes.DeadlineExceeded:
-				// Timeout. Not a fatal error, but note that it happened.
-				recordStat(it.ctx, ModAckTimeoutCount, 1)
-				return nil
-			default:
-				// This addresses an error where `context deadline exceeded` errors
-				// not captured by the previous case causes fatal errors.
-				// See https://github.com/googleapis/google-cloud-go/issues/3060
-				if err != nil && strings.Contains(err.Error(), "context deadline exceeded") {
-					recordStat(it.ctx, ModAckTimeoutCount, 1)
-					return nil
-				}
-				// TODO(b/226593754): by default, errors should not be fatal unless exactly once is enabled
-				// since modacks are "fire and forget". Once EOS feature is out, retry these errors
-				// if exactly-once is enabled, which can be determined from StreamingPull response.
-				return nil
+		addModAcks(toSend, deadlineSec)
+		// Use context.Background() as the call's context, not it.ctx. We don't
+		// want to cancel this RPC when the iterator is stopped.
+		cctx, cancel2 := context.WithTimeout(context.Background(), 60*time.Second)
+		defer cancel2()
+		err := it.subc.ModifyAckDeadline(cctx, &pb.ModifyAckDeadlineRequest{
+			Subscription:       it.subName,
+			AckDeadlineSeconds: deadlineSec,
+			AckIds:             toSend,
+		})
+		it.eoMu.RLock()
+		exactlyOnceDelivery := it.enableExactlyOnceDelivery
```

Comment:
This actually brings up an interesting point for me - if the EO semantics are enabled, and we receive messages and they're queued up for acks, and then EO is disabled... do we do the existing messages using the EO semantics, or switch immediately?

Reply:
Perhaps @maheshgattani can comment on this. I don't think switching exactly once delivery on/off is a common use case, so either behavior should be ok here. It's significantly easier to keep track of this through the actual on/off status of exactly once (through StreamingPullResponse) rather than keeping track of exactly once for every message.